Formula/deno.rb

```diff
@@ -0,0 +1,61 @@
+class Deno < Formula
+  desc "Command-line JavaScript / TypeScript engine"
+  homepage "https://deno.land/"
+  url "https://github.com/denoland/deno.git",
+    :tag      => "v0.6.0",
+    :revision => "22feb74ba12215597416b5531f8a557302283e79"
+
+  depends_on "llvm" => :build
+  depends_on "ninja" => :build
+  depends_on "node" => :build
+  depends_on "python@2" => :build # GN require Python 2.7+
+  depends_on "rust" => :build
+
+  # https://bugs.chromium.org/p/chromium/issues/detail?id=620127
+  depends_on :macos => :el_capitan
+
+  resource "gn" do
+    url "https://gn.googlesource.com/gn.git",
+      :revision => "64b846c96daeb3eaf08e26d8a84d8451c6cb712b"
+  end
+
+  def install
+    # Build gn from source and add it to the PATH
+    (buildpath/"gn").install resource("gn")
+    cd "gn" do
+      system "python", "build/gen.py"
+      system "ninja", "-C", "out/", "gn"
+    end
+    ENV.prepend_path "PATH", buildpath/"gn/out"
+
+    # change the path of gn and ninja to the homebrew provided ones (we don't do binary downloads here)
+    inreplace "tools/third_party.py" do |s|
+      s.gsub! "gn_path = tp(depot_tools_path, \"gn\")", "gn_path = \"#{buildpath}/gn/out/gn\""
+      s.gsub! "ninja_path = tp(depot_tools_path, \"ninja\")", "ninja_path = \"#{Formula["ninja"].bin}/ninja\""
```

Comment:
Sorry I'm late to the party here... but can someone catch me up:

Has this been reported/patched upstream? I.e., a way to avoid binary downloads?

In my mind, this is the only remaining blocker. Since this is patching the build, a `patch do` block is preferable to `inreplace` here. It should use the `url` and `sha256` form of `patch do` and have a comment explaining why the patch is needed, when it should be removed, and linking to upstream issue reports and/or PRs. Ideally the patch should be at homebrew/formula-patches (you can submit a PR) but if it's the .patch url of an upstream PR that's acceptable too.

Once this is done, squashing and rebasing ontop of the latest homebrew-core:master would also be greatly appreciated. Thanks!

Reply:
Yes, these are no longer being downloaded, ninja is installed via brew and gn is built from source

Reply:
@ry does --no-binary-downloads do the equivalent here? (Cam we replace the inreplace with that?)

Reply:
@zbeekman
> Has this been reported/patched upstream? I.e., a way to avoid binary downloads?

Upstream (similar to `v8` itself) only supports building with the binary devtools (`ninja`, `gn`, `clang`) downloaded from Googles `depot_tools`. This `inreplace` is not about preventing this binary download (we don't use the tooling which would cause the binary download here anymore, but instead call `ninja` and `gn` directly). It's about telling some part of the build (invoked by `gn`)  where the these tool are located, because some of them would just fail if they can't find `gn` at the well known binary `depot_tools` download path (even if a working version of `gn` is in the `PATH`, but the just don't even look for it there).

> In my mind, this is the only remaining blocker. Since this is patching the build, a `patch do` block is preferable to `inreplace` here. It should use the `url` and `sha256` form of `patch do` and have a comment explaining why the patch is needed, when it should be removed, and linking to upstream issue reports and/or PRs. Ideally the patch should be at homebrew/formula-patches (you can submit a PR) but if it's the .patch url of an upstream PR that's acceptable too.

Of cause we could create a patch for this. Alternatively working with even more symlinks for the `deno` build could work too.

> Once this is done, squashing and rebasing ontop of the latest homebrew-core:master would also be greatly appreciated. Thanks!

Sure

@hayd
> does --no-binary-downloads do the equivalent here? (Can we replace the inreplace with that?)
The inreplace is needed during the build so that scrips called by `gn` can find the correct path to `gn` and don't just look in the binary `depot_tools` location and fail because they can't find `gn`there. Maybe even more symlinks would fix this?
Also ` --no-binary-downloads` is incomplete, because it would still download `sccache`. Also it doesn't change anything about the assumption, where `gn` and `ninja` should be located (inside subdirectories of `depot_tools`).

Reply:
Thanks, I will push that feedback upstream!

Reply:
@hayd Thanks. Just to clarify, these are the 2 things witch are blocking us from using `setup.py` and `build.py` directly:
* make `setup.py`'s `--no-binary-downloads` option skip the download of `sccache` too
* make `third_party.py`'s `gn_path` and `ninja_path` (used by a lot of build tooling) configurable. I'd prefer using another gn-argument here (because we have to set `clang_base_pat` anyway), but I guess a simple environment variable would be ok too.

Reply:
> It's about telling some part of the build (invoked by `gn`) where these tool are located, because some of them would just fail if they can't find `gn` at the well known binary `depot_tools` download path

IMO, this is something upstream should be doing.

> Thanks, I will push that feedback upstream!

👍 

> @hayd Thanks. Just to clarify, these are the 2 things witch are blocking us from using `setup.py` and `build.py` directly...

Exactly!

Reply:
@chrmoritz Actually you don't need to call setup.py at all. Just call gn and ninja directly:
```
mkdir -p target/release
gn gen target/release
ninja -C target/release deno
```

Reply:
Not 100% sure, but @ry, you may be muddying the waters here. It looks like gn and ninja are the only build tools being called for the main formula. (Maybe you're looking at an outdated diff..., like the one in this thread.)

Reply:
@ry We are already calling `gn` and `ninja` directly. The problem is, that `gn` is spawning a script at some point, which fails if the `gn_path` in `third_party.py` isn't valid.

```
[474/638] python ../../tools/write_gn_args.py gen/cli/gn_args.txt
FAILED: gen/cli/gn_args.txt 
python ../../tools/write_gn_args.py gen/cli/gn_args.txt
gn.py: Could not find gn executable at: /private/tmp/deno-20190523-89403-1umch2/buildtools/mac/gn
Traceback (most recent call last):
  File "../../tools/write_gn_args.py", line 14, in <module>
    env=third_party.google_env())
  File "/private/tmp/deno-20190523-89403-1umch2/tools/util.py", line 65, in run_output
    return subprocess.check_output(args, cwd=cwd, env=env, shell=shell)
  File "/usr/local/Cellar/python@2/2.7.16/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 223, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
subprocess.CalledProcessError: Command '['/private/tmp/deno-20190523-89403-1umch2/third_party/depot_tools/gn', 'args', '/private/tmp/deno-20190523-89403-1umch2/target/release', '--list', '--short', '--overrides-only']' returned non-zero exit status 2
```
This is with calling `gn` directly, but `gn`only being in the users `PATH` and not located in `./buildtools/mac => ./third_party/v8/buildtools/mac`.

Reply:
@hayd Here is a working formula to showcase how the build could look like (using `setup.py` and `build.py`) if we get upstream support for setting `gn_path` and `ninja_path` through environment variables: https://gist.github.com/chrmoritz/3dfc9fa7acc720cf5e1289e51662f69c
(also it simply disables the `sccache` download, however this should probably be conditional based on the `--no-binary-downloads` argument)